Collapsible Column Groups Topic (#1691)
IgniteUI/igniteui-docfx: en/components/grids_templates/collapsible_column_groups.md

---
title: Angular Collapsible Column Groups | Ignite UI for Angular | Infragistics
_description: Collapsible multi-column headers make it possible to collapse some subset of the nested columns under the current one and to show some nested headers, which will give you a shorten informtion for example.
_keywords: collpasible column headers, ignite ui for angular, infragistics
---

### Grid Collapsible Column Groups Overview

> Multi-column headers allow you to have multiple levels of nested columns and column groups, also they provide you the ability to mark each column group as **collapsible**. **Collapsible multi-column headers** make it possible to collapse some nested headers under the current one and show some nested headers, which will give you a shorten informtion for example.

Reviewer: ```suggestion
**Collapsible multi-column headers** make it possible to collapse/expand, e.g. to show and hide the nested headers under the current one, which will give you a shortened/summarized information for example.
```

</div>
<div>
<button data-localize="stackblitz" disabled class="stackblitz-btn" data-iframe-id="hierarchical-grid-collapsible-column-groups-iframe" data-demos-base-url="{environment:demosBaseUrl}">view on stackblitz</button>
</div>
<div class="divider--half"></div>
}

### Setup

To get started with the IgxGrid and the **Collapsible multi-column headers** , first you need to install Ignite UI for Angular by typing the following command:

```cmd
ng add igniteui-angular
```
For a complete introduction to the Ignite UI for Angular, read the [*getting started*](general/getting_started.md) topic.

The next step is to import the @@if (igxName === 'IgxGrid') {`IgxGridModule`} @@if (igxName === 'IgxTreeGrid') {`IgxTreeGridModule`} @@if (igxName === 'IgxHierarchicalGrid') {`IgxHierarchicalGridModule`} in the app.module.ts file. Also we strongly suggest you to take a brief look at [*multi-column groups*](./multi_column_headers.md) topic, to see more detailed information on how to setup the column groups in your grid.

### Usage

> *Collapsible Column Groups* is a part of the multi-column headers feature which provides a way to collapse/expand a column group to a smaller set of data. When a column group is collapsed, a subset of the columns will be shown to the end-user and the other child columns of the group are going to be hidden. Each collapsed/expanded column can be bound to the data source of the grid, or it may be unbound, thus calculated.

Reviewer: ```suggestion
a subset of the columns will be shown to the end-user and the other child columns of the group will hide. Each collapsed/expanded column can be bound to the grid data source, or it may be unbound, thus calculated.
```

> In order to define a column group as `collapsible`, you need to set the property to `[collapsible]="true"` and also keep in mind that you need to define the property `visibleWhenCollapse` to at least two child columns. You need to set at least one column that will be visible when the parent column is collapsed (`[visibleWhenCollapse]="true"`) and at least one column should be set to be visible when the parent is expanded (`[visibleWhenCollapse]="false"`), otherwise the **collapsible fuctionallity will be disabled**. If you don't specify the property `visibleWhenCollapse` for some of the child columns - means that that column will be always visible no matter whether the parent state is expanded or collapsed.

Reviewer: ```suggestion
In order to define a column group as `collapsible`, you need to set the property to `[collapsible]="true"` and also keep in mind that you need to define the property `visibleWhenCollapse` to at least two child columns: at least one column must be visible when the group is collapsed (`[visibleWhenCollapse]="true"`) and at least one column must be hidden when the group is expanded (`[visibleWhenCollapse]="false"`), otherwise the **collapsible functionality will be disabled**. If  `visibleWhenCollapse` is not specified for some of the child columns, then this column will be always visible no matter whether the parent state is expanded or collapsed.
```

    <!--The three columns below will be visible when its parent is expanded-->
    <igx-column field="CustomerID" header="Customer ID" [dataType]="'string'" [visibleWhenCollapsed]="false"></igx-column>
    <igx-column field="FirstName" header="First Name" [dataType]="'string'" [visibleWhenCollapsed]="false">
    </igx-column>
    <igx-column field="LastName" header="Last Name" [dataType]="'string'" [visibleWhenCollapsed]="false">
    </igx-column>
    <igx-column-group header="Customer Address"> <!--This column visibility will not be changed based on parent expand/collapsed state-->
        <igx-column field="Country" header="Country" [dataType]="'string'" [sortable]="true">
        </igx-column>
        <igx-column field="City" header="City" [dataType]="'string'" [sortable]="true">
        </igx-column>
    </igx-column-group>
</igx-column-group>
```

And now let's sum up: every child column has tree states:
-	Can be always visible, no matter expand state of its parent;
-	Can be visible, when its parent is expanded;
-	Can be visible, when its parent is collapsed;

> The initial state of the column group which is specified as collapsible is `[expanded]="true"`. But you can easy change this behavour by setting the property `[expand]="false"`.

Reviewer: easily

### Expand/Collapse indicator template

Default expand indicator for the igxGrid is the following:

 <img src="../../images/general/expand_indicator.png" style="width: 450px; height: 130px"/>

Default collapse indicator for the igxGrid is the following:

<img src="../../images/general/collapsed_indicator.png" style="width: 400px; height: 130px"/>

> Also if you need to change the default expand/collapse indicator we introduce you two easy way to do it.

Reviewer: two easy ways

You can define custom expand/collapse template and provide it to each of the collapsible column groups using **collapsibleIndicatorTemplate** input property. Check the markup below:

```html
<ng-template #indTemplate let-column="column">
    <igx-icon [attr.draggable]="false" >{{column.expanded ? 'remove' : 'add'}} </igx-icon>
</ng-template>
<igx-column-group header="Customer Information" [collapsible]="true" [collapsibleIndicatorTemplate]="indTemplate">
    <igx-column field="CustomerName" header="Fullname" [dataType]="'string'" [visibleWhenCollapsed]="true"></igx-column>
    <igx-column field="CustomerID" header="Customer ID" [dataType]="'string'" [visibleWhenCollapsed]="false"></igx-column>
    <igx-column-group header="Customer Address" [collapsible]="true" [collapsibleIndicatorTemplate]="indTemplate">
        <igx-column field="Country" header="Country" [dataType]="'string'" [sortable]="true" [visibleWhenCollapsed]="true"></igx-column>
        <igx-column field="City" header="City" [dataType]="'string'" [sortable]="true" [visibleWhenCollapsed]="false"></igx-column>
    </igx-column-group>
</igx-column-group>
```
##### Using igxCollapsibleIndicator directive

> Another way to achieve this behavior is to use the igxCollapsibleIndicator directive like in the example below:

Reviewer: as shown in the example below:
